docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerSlave.java

```diff
@@ -135,43 +147,101 @@ private void slaveShutdown(TaskListener listener) throws DockerException, IOExce
 
-        // Tag it with the jenkins name
-        addJenkinsAction(tag_image);
+        // Iterate list of tags, and tag appropriately
+        for(String tag : imageTags) {
+            // Clean up the image tag
+            String cleanTag = cleanImageTag(tag);
 
-        // SHould we add additional tags?
-        try
-        {
-            String tagToken = getAdditionalTag(listener);
+            // Skip empty tags
+            if(Strings.isNullOrEmpty(cleanTag)){
+                continue;
+            }
+            LOGGER.log(Level.INFO, "Tagging with: " + cleanTag);
 
-            if( !Strings.isNullOrEmpty(tagToken) ) {
-                // ?? client.image(tag_image).tag(tagToken, false);
-                client.tagImageCmd(tag_image,null,tagToken).exec();
-                addJenkinsAction(tagToken);
+            // Tag the image!
+            try {
+                client.tagImageCmd(imageId, repositoryName, cleanTag)
+                    .withForce()
+                    .exec();
+                // addJenkinsAction(cleanTag);
+            } catch(Exception ex) {
+                LOGGER.log(Level.SEVERE, "Could not add tag: " + cleanTag, ex);
+            }
 
-                if( getJobProperty().pushOnSuccess ) {
-                    client.pushImageCmd(tagToken).exec();
+            // Push the image to the registry with all of the tags
+            if(getJobProperty().isPushOnSuccess()) {
+                try {
+                    client.pushImageCmd(repositoryName)
+                        .withTag(cleanTag)
+                        .exec();
+                } catch (NotFoundException ex) {
+                    // This should NEVER happen but just in case.
+                    LOGGER.log(Level.SEVERE, "Couldnt find " + repositoryName + ":" +cleanTag + ". Not pushing.");
                 }
             }
         }
-        catch(Exception ex) {
-            LOGGER.log(Level.SEVERE, "Could not add additional tags");
-        }
 
-        if( getJobProperty().cleanImages ) {
+        // Add a Jenkins 'Action' for the newly committed image
+        addJenkinsAction(imageId);
 
-            client.removeImageCmd(tag_image)
+        if( getJobProperty().cleanImages ) {
+            client.removeImageCmd(imageId)
                 .withForce()
                 .exec();
         }
+    }
 
+    /* Image names are made up of: [registry.domain/][username/]repository:tag
+     * Registry should be a valid domain, with at least one dot (.)
+     * Namespace and repository should match [a-z0-9-_.]
+     * Tags should match [A-Za-z0-9_.-]
+     * TODO: Match each part of the full repository name.
+     */
+    private static String cleanImageRepositoryName(String repositoryName) {
+        return repositoryName.toLowerCase()
+                             .replaceAll("\\s", "_")
+                             .replaceAll("[^/a-z0-9-_.]", "");
+    }
+
+    private static String cleanImageTag(String imageTag) {
```

Comment:
imho better call function `filterTagName()` because cleanImageTag looks for me like remove tags from real image.